core/src/main/scala/com/gu/tableversions/core/TableVersions.scala

```diff
@@ -13,39 +17,93 @@ trait TableVersions[F[_]] {
+      updatesForCurrentVersion = matchingUpdates._1 ++ matchingUpdates._2.take(1)
+      operations = updatesForCurrentVersion.flatMap(_.operations)
+    } yield
+      if (isSnapshotTable(operations))
+        latestSnapshotTableVersion(operations)
+      else
+        applyPartitionUpdates(PartitionedTableVersion(Map.empty))(operations)
+
+  /** Return the history of table updates, most recent first. */
+  def updates(table: TableName)(implicit F: Sync[F]): F[List[TableUpdateHeader]] =
+    tableState(table).map(_.updates.map(_.header).reverse)
 
   /**
     * Update partition versions to the given versions.
-    * This performs no checking if data has been written to the associated paths but it will verify that these versions
-    * 1) haven't been committed before and 2) these are the next versions to be committed for each of the partitions.
     */
-  def commit(table: TableName, update: TableVersions.TableUpdate): F[CommitResult]
+  def commit(table: TableName, update: TableVersions.TableUpdate): F[Unit]
+
+  /**
+    * Set the current version of a table to refer to an existing version.
+    */
+  def setCurrentVersion(table: TableName, id: TableVersions.CommitId): F[Unit]
+
+  //
+  // Internal operations to be provided by implementations
+  //
+
+  /**
+    * Produce description of the current state of table.
+    */
+  protected def tableState(table: TableName): F[TableState]
+
+  /**
+    * Handle initialisation of a new table if it doesn't exist already.
+    */
+  protected def handleInit(table: TableName)(newTableState: => TableState): F[Unit]
 
 }
 
 object TableVersions {
 
-  /** A collection of updates to partitions to be applied and tracked as a single atomic change. */
-  final case class TableUpdate(
+  final case class TableUpdateHeader(
+      id: CommitId,
       userId: UserId,
       message: UpdateMessage,
-      timestamp: Instant,
-      operations: List[TableOperation])
+      timestamp: Instant
+  )
 
-  final case class UpdateMessage(content: String) extends AnyVal
+  object TableUpdateHeader {
+
+    def apply(userId: UserId, message: UpdateMessage, timestamp: Instant): TableUpdateHeader =
+      TableUpdateHeader(createId(), userId, message, timestamp)
+
+    private def createId(): CommitId = CommitId(UUID.randomUUID().toString)
+  }
+
+  final case class CommitId(id: String) extends AnyVal
 
   final case class UserId(value: String) extends AnyVal
 
-  /** Result type for commit operation */
-  sealed trait CommitResult
+  final case class UpdateMessage(content: String) extends AnyVal
 
-  object CommitResult {
-    case object SuccessfulCommit extends CommitResult
-    final case class InvalidCommit(invalidPartitions: Map[Partition, ErrorMessage]) extends CommitResult
+  /** A collection of updates to partitions to be applied and tracked as a single atomic change. */
+  final case class TableUpdate(header: TableUpdateHeader, operations: List[TableOperation])
```

Comment:
The suffix `Header` induced me in error thinking it had something to do with a table's header, or maybe schema.
How about `TableOperationMetadata`, so it semantically refers to the name of the `operations` type ?

Reply:
True, I just found that 'metadata' is quite heavily used, e.g. `TableMetadata` (albeit not in this repo) along with "metastore".

Not sure about `TableOperationMetadata` as it's metadata pertaining to the table update, not the operations... how about `TableUpdateMetadata`?

Reply:
Works for me +1